ffi/src/ir.rs

```diff
@@ -0,0 +1,249 @@
+use alloc::{boxed::Box, string::String, vec::Vec};
+use core::mem::ManuallyDrop;
+
+use crate::{repr_c::CType, slice::SliceRef, FfiReturn, ReprC, Result};
+
+/// Trait for `#[repr(transparent)]` structs to convert between [`Self`] and [`Self::Inner`]
+///
+/// # Safety
+/// `Self` and `Self::Inner` must have the same memory layout, the same set of valid values,
+/// and calling convention
+pub unsafe trait Transparent: Sized {
+    /// Non-ZST field of transparent struct
+    type Inner: Ir;
+
+    /// Convert transparent struct into its non-ZST field
+    // NOTE: Deriving `into_inner` is always valid because T::Inner is a superset of T
+    fn into_inner(outer: Self) -> Self::Inner;
+
+    /// Recover transparent struct from its non-ZST field
+    // FIXME: We should force the transparent type to have `from_inner` and derive implementation
+    // only if requested specifically by the user (via macro attribute). The reason for this is
+    // that even though two types are equal on the byte level, they may have different semantics
+    // that cannot be proven when deriving `from_inner` (e.g. `NonZeroU8` is a wrapper for `u8`)
+    // Deriving implementation of this method cannot be guaranteed to be correct
+    unsafe fn from_inner(inner: Self::Inner) -> Self;
```

Comment:
In this case should't this function be `try_from_inner`?

Reply:
true, I resorted to returning `Option<Self>`